apps/roam/src/utils/conceptConversion.ts

```diff
@@ -0,0 +1,184 @@
+  const created = new Date(created_t).toISOString();
+  const last_modified = new Date(last_modified_t).toISOString();
+  return {
+    author_uid,
+    created,
+    last_modified,
+    page_uid
+  }
+}
+
+
+export function discourseNodeSchemaToLocalConcept(
+  context: SupabaseContext,
+  node: DiscourseNode, // contains type, uid, text
+): LocalConceptDataInput {
+  const extra_data = getNodeExtraData(node.type);
+  const title_parts = node.text.split('/');
+  return {
+    space_id: context.spaceId,
+    author_local_id: extra_data.author_uid,
+    created: extra_data.created,
+    last_modified: extra_data.last_modified,
+    name: title_parts[title_parts.length - 1],
+    represented_by_local_id: node.type,
+    is_schema: true,
+  };
+}
+
+
+export function discourseNodeBlockToLocalConcept(
+  context: SupabaseContext,
+  {
+    node_uid,
+    schema_uid,
+    text
+  }: {
+    node_uid: string,
+    schema_uid: string,
+    text: string
+  }
+): LocalConceptDataInput {
+  const extra_data = getNodeExtraData(node_uid);
+  return {
+    space_id: context.spaceId,
+    author_local_id: extra_data.author_uid,
+    created: extra_data.created,
+    last_modified: extra_data.last_modified,
+    name: text,
+    represented_by_local_id: node_uid,
+    schema_represented_by_local_id: schema_uid,
+    is_schema: false,
+  };
+
+}
+
+const STANDARD_ROLES = ["source", "target"];
+
+export function discourseRelationSchemaToLocalConcept(
+  context: SupabaseContext,
+  relation: DiscourseRelation, // contains type, uid, text
+): LocalConceptDataInput {
+  const extra_data = getNodeExtraData(relation.id);
+  return {
+    space_id: context.spaceId,
+    represented_by_local_id: relation.id,
+    author_local_id: extra_data.author_uid,
+    created: extra_data.created,
+    last_modified: extra_data.last_modified,
+    name: `${relation.id}-${relation.label}`,  // the label is not unique
+    is_schema: true,
+    local_reference_content: Object.fromEntries(
+      Object.entries(relation)
+        .filter(([key, v]) => STANDARD_ROLES.includes(key))
+      ) as { [key: string]: string },
+    literal_content: {
+      roles: STANDARD_ROLES,
+      label: relation.label,
+      complement: relation.complement,
+      representation: relation.triples.map(t=>t[0]),
+    },
+  };
+}
+
+export function discourseRelationDataToLocalConcept(
+  context: SupabaseContext,
+  relationSchemaUid: string,
+  relationNodes: { [role: string]: string },
+): LocalConceptDataInput {
+  const roamRelation = getDiscourseRelations().find(r => r.id === relationSchemaUid);
+  if (roamRelation === undefined) {
+    throw new Error(`Invalid roam relation id ${relationSchemaUid}`)
+  }
+  const relation = discourseRelationSchemaToLocalConcept(context, roamRelation);
+  const lit_content = ((relation.literal_content)? relation.literal_content: {}) as unknown as {[key: string]: any};
```

Comment:
Why is this assertion necessary?  Can we not achieve this with types?

Reply:
The type of `literal_content` is `Json | undefined`. I have to assert typing to tell it about the expected Json structure, otherwise I run in even more trouble in the next line.
Do you think that would warrant a comment? (I'm really confused as to what should be considered obvious or commendable.)